Fehler wegen immutable username bei Klasse User

Beim Anlegen von Test-User über Flyway wird eine Exception geworfen (s.u.).
Grund ist die Annotation `@NaturalId` beim Attribut `username`in der Entität `User`. 
Die Annotation sorgt dafür das das Feld "immutable" ist.


`Caused by: org.springframework.orm.jpa.JpaSystemException: An immutable natural identifier of entity de.muenchen.oss.wahllokalsystem.authservice.domain.User was altered from `user` to `ENCRYPTED:4YH1c_dIti6yTl2-01G3hg==`
	at org.springframework.orm.jpa.vendor.HibernateJpaDialect.convertHibernateAccessException(HibernateJpaDialect.java:341)
	at org.springframework.orm.jpa.vendor.HibernateJpaDialect.translateExceptionIfPossible(HibernateJpaDialect.java:241)
	at org.springframework.orm.jpa.JpaTransactionManager.doCommit(JpaTransactionManager.java:566)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.processCommit(AbstractPlatformTransactionManager.java:795)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.commit(AbstractPlatformTransactionManager.java:758)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.commitTransactionAfterReturning(TransactionAspectSupport.java:663)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:413)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:119)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184)
	at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:138)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184)
	at org.springframework.data.jpa.repository.support.CrudMethodMetadataPostProcessor$CrudMethodMetadataPopulatingMethodInterceptor.invoke(CrudMethodMetadataPostProcessor.java:165)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184)
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:97)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:223)
	at de.muenchen.oss.wahllokalsystem.authservice.domain.$Proxy185.saveAll(Unknown Source)
	at de.muenchen.oss.wahllokalsystem.authservice.domain.UserRepositoryImpl.encryptUsernames(UserRepositoryImpl.java:97)
	at de.muenchen.oss.wahllokalsystem.authservice.domain.UserRepositoryImpl.onInit(UserRepositoryImpl.java:36)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor$LifecycleMethod.invoke(InitDestroyAnnotationBeanPostProcessor.java:457)
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor$LifecycleMetadata.invokeInitMethods(InitDestroyAnnotationBeanPostProcessor.java:401)
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor.postProcessBeforeInitialization(InitDestroyAnnotationBeanPostProcessor.java:219)
	... 18 common frames omitted
Caused by: org.hibernate.HibernateException: An immutable natural identifier of entity de.muenchen.oss.wahllokalsystem.authservice.domain.User was altered from `user` to `ENCRYPTED:4YH1c_dIti6yTl2-01G3hg==`
	at org.hibernate.metamodel.mapping.internal.SimpleNaturalIdMapping.verifyFlushState(SimpleNaturalIdMapping.java:94)
	at org.hibernate.event.internal.DefaultFlushEntityEventListener.checkNaturalId(DefaultFlushEntityEventListener.java:113)
	at org.hibernate.event.internal.DefaultFlushEntityEventListener.getValues(DefaultFlushEntityEventListener.java:185)
	at org.hibernate.event.internal.DefaultFlushEntityEventListener.onFlushEntity(DefaultFlushEntityEventListener.java:141)
	at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:127)
	at org.hibernate.event.internal.AbstractFlushingEventListener.flushEntities(AbstractFlushingEventListener.java:242)
	at org.hibernate.event.internal.AbstractFlushingEventListener.flushEverythingToExecutions(AbstractFlushingEventListener.java:88)
	at org.hibernate.event.internal.AbstractFlushingEventListener.flushEverythingToExecutions(AbstractFlushingEventListener.java:82)
	at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:40)
	at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:127)
	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1425)
	at org.hibernate.internal.SessionImpl.managedFlush(SessionImpl.java:487)
	at org.hibernate.internal.SessionImpl.flushBeforeTransactionCompletion(SessionImpl.java:2324)
	at org.hibernate.internal.SessionImpl.beforeTransactionCompletion(SessionImpl.java:1981)
	at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.beforeTransactionCompletion(JdbcCoordinatorImpl.java:439)
	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl.beforeCompletionCallback(JdbcResourceLocalTransactionCoordinatorImpl.java:169)
	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl$TransactionDriverControlImpl.commit(JdbcResourceLocalTransactionCoordinatorImpl.java:267)
	at org.hibernate.engine.transaction.internal.TransactionImpl.commit(TransactionImpl.java:101)
	at org.springframework.orm.jpa.JpaTransactionManager.doCommit(JpaTransactionManager.java:562)`

--- a/wls-auth-service/src/main/java/de/muenchen/oss/wahllokalsystem/authservice/domain/User.java
+++ b/wls-auth-service/src/main/java/de/muenchen/oss/wahllokalsystem/authservice/domain/User.java
@@ -19,7 +19,6 @@ import lombok.EqualsAndHashCode;
 import lombok.NoArgsConstructor;
 import lombok.ToString;
 import lombok.val;
-import org.hibernate.annotations.NaturalId;
 
 @Entity
 @Table(name = "Wlsuser") //user as table name is already in use by h2
@@ -39,7 +38,6 @@ public class User extends BaseEntity {
         return copy;
     }
 
-    @NaturalId
     @NotNull
     @Size(min = 1)
     @ToString.Include

--- a/wls-auth-service/src/main/resources/application-test.yml
+++ b/wls-auth-service/src/main/resources/application-test.yml
@@ -1,15 +1,13 @@
 spring:
 
+  flyway:
+    enabled: true
   # Spring JPA
   h2.console.enabled: true
   jpa:
     database: H2
     hibernate:
-      # always drop and create the db should be the best
-      # configuration for local (development) mode. this
-      # is also the default, that spring offers by convention.
-      # but here explicite:
-      ddl-auto: create-drop
+      ddl-auto: validate
       naming.physical-strategy: org.hibernate.boot.model.naming.PhysicalNamingStrategyStandardImpl
     # Logging for database operation
     show-sql: true

--- a/wls-auth-service/src/test/java/de/muenchen/oss/wahllokalsystem/authservice/domain/UserRepositoryImplIntegrationTest.java
+++ b/wls-auth-service/src/test/java/de/muenchen/oss/wahllokalsystem/authservice/domain/UserRepositoryImplIntegrationTest.java
@@ -14,6 +14,7 @@ import org.junit.jupiter.api.Nested;
 import org.junit.jupiter.api.Test;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.dao.DataIntegrityViolationException;
 import org.springframework.test.context.ActiveProfiles;
 import org.springframework.transaction.support.TransactionTemplate;
 
@@ -129,6 +130,19 @@ class UserRepositoryImplIntegrationTest {
 
             Assertions.assertThat(savedUser.getUsername()).isEqualTo(USERNAME_UNENCRYPTED);
         }
+
+        @Test
+        void should_throwException_whenUserWithUsernameAlreadyExists() {
+            val userToSave = new User();
+            userToSave.setUsername(USERNAME_UNENCRYPTED);
+
+            transactionTemplate.execute(status -> userRepository.save(userToSave));
+
+            val userToSaveWithSameUsername = new User();
+            userToSaveWithSameUsername.setUsername(USERNAME_UNENCRYPTED);
+            Assertions.assertThatException().isThrownBy(() -> transactionTemplate.execute(status -> userRepository.save(userToSaveWithSameUsername)))
+                    .isInstanceOf(DataIntegrityViolationException.class);
+        }
     }
 
     @Nested
@@ -216,6 +230,42 @@ class UserRepositoryImplIntegrationTest {
             Assertions.assertThat(savedUsers).allSatisfy(user -> Assertions.assertThat(crudUserRepository.existsById(user.getId())).isFalse());
             Assertions.assertThat(authorityRepository.count()).isEqualTo(1);
             Assertions.assertThat(permissionRepository.count()).isEqualTo(3);
+        }
+    }
+
+    @Nested
+    class OnSchedule {
+
+        @Test
+        void should_encryptExistingUsers_when_usersExists() {
+            val wahltagID = "wahltagID";
+            val userToEncrypt = createUser(USERNAME_UNENCRYPTED, wahltagID);
+
+            val savedSavedUnencryptedUser = transactionTemplate.execute(status -> crudUserRepository.save(userToEncrypt));
+
+            transactionTemplate.executeWithoutResult(status -> userRepository.onSchedule());
+
+            val userAfterEncryption = crudUserRepository.findById(savedSavedUnencryptedUser.getId()).get();
+
+            Assertions.assertThat(userAfterEncryption.getUsername()).isEqualTo(USERNAME_ENCRYPTED);
+        }
+    }
+
+    @Nested
+    class OnInit {
+
+        @Test
+        void should_encryptExistingUsers_when_usersExists() {
+            val wahltagID = "wahltagID";
+            val userToEncrypt = createUser(USERNAME_UNENCRYPTED, wahltagID);
+
+            val savedSavedUnencryptedUser = transactionTemplate.execute(status -> crudUserRepository.save(userToEncrypt));
+
+            transactionTemplate.executeWithoutResult(status -> userRepository.onSchedule());
+
+            val userAfterEncryption = crudUserRepository.findById(savedSavedUnencryptedUser.getId()).get();
+
+            Assertions.assertThat(userAfterEncryption.getUsername()).isEqualTo(USERNAME_ENCRYPTED);
         }
     }
 

--- a/wls-auth-service/src/test/java/de/muenchen/oss/wahllokalsystem/authservice/domain/UserRepositoryImplIntegrationTest.java
+++ b/wls-auth-service/src/test/java/de/muenchen/oss/wahllokalsystem/authservice/domain/UserRepositoryImplIntegrationTest.java
@@ -171,6 +171,19 @@ class UserRepositoryImplIntegrationTest {
 
             Assertions.assertThat(savedUser).allSatisfy(user -> Assertions.assertThat(user.getUsername()).isEqualTo(USERNAME_UNENCRYPTED));
         }
+
+        @Test
+        void should_throwException_whenMultipleUsersHaveSameName() {
+            val user1 = new User();
+            user1.setUsername(USERNAME_UNENCRYPTED);
+            val user2 = new User();
+            user2.setUsername(USERNAME_UNENCRYPTED);
+
+            val userToSaveWithSameUsername = new User();
+            userToSaveWithSameUsername.setUsername(USERNAME_UNENCRYPTED);
+            Assertions.assertThatException().isThrownBy(() -> transactionTemplate.execute(status -> userRepository.saveAll(List.of(user1, user2))))
+                    .isInstanceOf(DataIntegrityViolationException.class);
+        }
     }
 
     @Nested

--- a/wls-auth-service/src/test/java/de/muenchen/oss/wahllokalsystem/authservice/domain/UserRepositoryImplIntegrationTest.java
+++ b/wls-auth-service/src/test/java/de/muenchen/oss/wahllokalsystem/authservice/domain/UserRepositoryImplIntegrationTest.java
@@ -274,7 +274,7 @@ class UserRepositoryImplIntegrationTest {
 
             val savedSavedUnencryptedUser = transactionTemplate.execute(status -> crudUserRepository.save(userToEncrypt));
 
-            transactionTemplate.executeWithoutResult(status -> userRepository.onSchedule());
+            transactionTemplate.executeWithoutResult(status -> userRepository.onInit());
 
             val userAfterEncryption = crudUserRepository.findById(savedSavedUnencryptedUser.getId()).get();
 

--- a/wls-auth-service/src/test/java/de/muenchen/oss/wahllokalsystem/authservice/domain/UserRepositoryImplIntegrationTest.java
+++ b/wls-auth-service/src/test/java/de/muenchen/oss/wahllokalsystem/authservice/domain/UserRepositoryImplIntegrationTest.java
@@ -132,7 +132,7 @@ class UserRepositoryImplIntegrationTest {
         }
 
         @Test
-        void should_throwException_whenUserWithUsernameAlreadyExists() {
+        void should_throwException_when_userWithUsernameAlreadyExists() {
             val userToSave = new User();
             userToSave.setUsername(USERNAME_UNENCRYPTED);
 
@@ -173,7 +173,7 @@ class UserRepositoryImplIntegrationTest {
         }
 
         @Test
-        void should_throwException_whenMultipleUsersHaveSameName() {
+        void should_throwException_when_multipleUsersHaveSameName() {
             val user1 = new User();
             user1.setUsername(USERNAME_UNENCRYPTED);
             val user2 = new User();

